packages/validator/src/services/block.ts

```diff
@@ -77,50 +82,109 @@ export class BlockProposingService {
       this.logger.debug("Producing block", debugLogCtx);
       this.metrics?.proposerStepCallProduceBlock.observe(this.clock.secFromSlot(slot));
 
-      const block = await this.produceBlock(slot, randaoReveal, graffiti).catch((e: Error) => {
-        this.metrics?.blockProposingErrors.inc({error: "produce"});
-        throw extendError(e, "Failed to produce block");
+      const expectedFeeRecipient = this.validatorStore.feeRecipientByValidatorPubkey.getOrDefault(pubkeyHex);
+      const block = await this.produceBlockWrapper(slot, randaoReveal, graffiti, expectedFeeRecipient).catch(
+        (e: Error) => {
+          this.metrics?.blockProposingErrors.inc({error: "produce"});
+          throw extendError(e, "Failed to produce block");
+        }
+      );
+
+      this.logger.debug("Produced block", {...debugLogCtx, ...block.debugLogCtx});
+      this.metrics?.blocksProduced.inc();
+
+      const signedBlock = await this.validatorStore.signBlock(block.type, pubkey, block.data, slot);
+
+      this.metrics?.proposerStepCallPublishBlock.observe(this.clock.secFromSlot(slot));
+
+      await this.publishBlockWrapper(block.type, signedBlock).catch((e: Error) => {
+        this.metrics?.blockProposingErrors.inc({error: "publish"});
+        throw extendError(e, "Failed to publish block");
       });
-      const blockFeeRecipient = (block.data as bellatrix.BeaconBlock).body.executionPayload?.feeRecipient;
+      this.logger.info("Published block", {...logCtx, graffiti, ...block.debugLogCtx});
+      this.metrics?.blocksPublished.inc();
+    } catch (e) {
+      this.logger.error("Error proposing block", logCtx, e as Error);
+    }
+  }
+
+  private publishBlockWrapper = async <T extends allForks.BlockType>(
+    type: T,
+    signedBlock: allForks.FullOrBlindedSignedBeaconBlock<T>
+  ): Promise<void> => {
+    switch (type) {
+      case allForks.BlockType.Blinded:
+        return this.api.beacon.publishBlindedBlock(signedBlock as bellatrix.SignedBlindedBeaconBlock);
+        break;
+      case allForks.BlockType.Full:
+      default:
+        return this.api.beacon.publishBlock(signedBlock);
+    }
+  };
+
+  private produceBlockWrapper = async (
+    slot: Slot,
+    randaoReveal: BLSSignature,
+    graffiti: string,
+    expectedFeeRecipient: string
+  ): Promise<WrappedProduceBlockType & {debugLogCtx: Record<string, string>}> => {
+    const blindedBlockPromise = this.opts.builder.enabled
+      ? this.api.validator.produceBlindedBlock(slot, randaoReveal, graffiti).catch((e: Error) => {
+          this.logger.error("Failed to produce builder block", {}, e as Error);
+          return null;
+        })
+      : null;
+
+    const fullBlockPromise = this.produceBlock(slot, randaoReveal, graffiti).catch((e: Error) => {
+      this.logger.error("Failed to produce builder block", {}, e as Error);
+      return null;
+    });
+
+    await Promise.all([blindedBlockPromise, fullBlockPromise]);
```

Comment:
Could you explain in great detail the strategy here? From what I can see you attempt to produce blocks and publish one prioritizing the blindedBlock.

But, what if the `produceBlindedBlock` take too long to resolve? You will publish the regular block very late.

Reply:
as discussed over discord, we can just go with this for now and try to come up with a better strategy.